registry/registry.go

```diff
@@ -2,68 +2,70 @@
 package registry
 
 import (
-	"context"
-	"encoding/base64"
-	"encoding/json"
-	"fmt"
-	"io/ioutil"
-	"net/http"
-	"net/http/cookiejar"
-	"sort"
-	"strconv"
-	"strings"
-	"time"
-
 	"github.com/go-kit/kit/log"
-	dockerregistry "github.com/heroku/docker-registry-client/registry"
-	"github.com/jonboulle/clockwork"
-	"golang.org/x/net/publicsuffix"
-
 	"github.com/weaveworks/flux"
-	fluxmetrics "github.com/weaveworks/flux/metrics"
+	"sort"
+	"time"
 )
 
 const (
-	dockerHubHost    = "index.docker.io"
-	dockerHubLibrary = "library"
-
 	requestTimeout = 10 * time.Second
 )
 
-type creds struct {
-	username, password string
+// Old registry interface. Used as an adaptor to the new class
+type Client interface {
+	GetRepository(repository string) ([]flux.ImageDescription, error)
 }
 
-// Credentials to a (Docker) registry.
-type Credentials struct {
-	m map[string]creds
+type registryAdapter struct {
+	r Registry
 }
 
-// Client is a handle to a bunch of registries.
-type Client interface {
-	GetRepository(repository string) ([]flux.ImageDescription, error)
+func NewClient(r Registry) Client {
+	return &registryAdapter{
+		r: r,
+	}
 }
 
-// client is a handle to a registry.
-type client struct {
-	Credentials Credentials
-	Logger      log.Logger
-	Metrics     Metrics
+func (a *registryAdapter) GetRepository(repository string) (res []flux.ImageDescription, err error) {
+	img, err := ParseImage(repository, nil)
+	if err != nil {
+		return
+	}
+	images, err := a.r.GetRepository(img)
+	if err != nil {
+		return
+	}
+	res = make([]flux.ImageDescription, len(images))
+	for i, im := range images {
+		res[i] = flux.ImageDescription{
+			ID:        flux.ParseImageID(im.FQN()),
+			CreatedAt: im.CreatedAt(),
+		}
+	}
+	return
 }
 
-// NewClient creates a new registry client, to use when fetching repositories.
-func NewClient(c Credentials, l log.Logger, m Metrics) Client {
-	return &client{
-		Credentials: c,
-		Logger:      l,
-		Metrics:     m,
-	}
+// New registry interface
+type Registry interface {
+	GetRepository(repository Image) ([]Image, error)
+	GetImage(repository Image) (Image, error)
```

Comment:
😬  Rather not string. No constraints on a string, so you'd get people using `host/namespace/repo` or `namespace/repo` or `namespace/repo:tag`. Better to be explicit with a type.

And although rather redundant (all the info is in an image), I like the naming of Repository. It reads nicer.